File: R/concordance.R
concordance <- function(x, y) {

  n <- length(x)
  stopifnot(length(y) == n)
  if (n == 0L) return(integer(0L))
  if (n == 1L) return(0L)
  if (n <= 130) concordance_naive(x, y)
  concordance_fenwick(x, y)

}

concordance_naive <- function(x, y) {
  dx <- sign(outer(x, x, `-`))
  dy <- sign(outer(y, y, `-`))
  .rowSums(dx * dy, nrow(dx), ncol(dx))
}

concordance_fenwick <- function(x, y) {

  n <- length(x)

  # 1) order by x (stable), keep x-ordered copies
  ord <- order(x, y, method = "radix")
  xo <- x[ord]
  yo <- y[ord]

  # 2) compute group runs for equal x (use rle to avoid building big lists)
  if (anyDuplicated(xo)) {
    rle_x <- rle(xo)
    group_lengths <- rle_x$lengths
    group_starts <- cumsum(c(1L, head(group_lengths, -1L)))
    group_ends   <- cumsum(group_lengths)
    G <- length(group_lengths)  # number of groups
  } else {
    # fast path
    rle_x <- rle(xo)
    group_lengths <- rep.int(1L, n)
    group_starts <- 1:n
    group_ends   <- 1:n
    G <- n  # number of groups
  }

  # 3) compress y to ranks 1..m (strict ordering of unique y's)
  uniq_y <- sort(unique(yo))
  ry <- match(yo, uniq_y)      # integer vector 1..m
  m <- length(uniq_y)

  # allocate bit and result vectors (integers)
  bit <- integer(m)            # 1-indexed BIT (positions 1..m)
  less_right    <- integer(n)
  greater_right <- integer(n)
  # originally there were separate vectors, but we can reuse some memory
  # less_left     <- integer(n)
  # greater_left  <- integer(n)

  # precompute the results of bitwAnd so we can just do lookup
  # lowbits <- (1:m) - ((1:m) & ((1:m) - 1L))
  # lowbits <- (1:m) - bitwAnd((1:m), ((1:m) - 1L))
  #lowbits <- vapply(1:m, \(i) bitwAnd(i, -i), 0L)# -> lowbits
  lowbits <- bitwAnd(1:m, -1:-m)

  # Helper: inline bit_sum and bit_add are implemented as loops below
  # -------- Right sweep (groups processed from right to left) ----------
  total_seen <- 0L
  for (g in G:1L) {
    i1 <- group_starts[g]
    i2 <- group_ends[g]
    # query each index in this group (do NOT add them yet)
    for (i in i1:i2) {
      r <- ry[i]
      # less_right: bit_sum(r-1)
      j <- r - 1L
      s_less <- 0L
      while (j > 0L) {
        s_less <- s_less + bit[j]
        j <- j - lowbits[j] #bitwAnd(j, -j)
      }
      # leq_count: bit_sum(r)
      j <- r
      s_leq <- 0L
      while (j > 0L) {
        s_leq <- s_leq + bit[j]
        j <- j - lowbits[j] #bitwAnd(j, -j)
      }
      less_right[i] <- s_less
      greater_right[i] <- total_seen - s_leq
    }
    # now add this entire group into the BIT (so earlier groups see them)
    for (i in i1:i2) {
      r <- ry[i]
      j <- r
      while (j <= m) {
        bit[j] <- bit[j] + 1L
        j <- j + lowbits[j] #bitwAnd(j, -j)
      }
      total_seen <- total_seen + 1L
    }
  }
  out_xorder <- (greater_right - less_right)  # partial result
  # reset for left sweep, these are now less_left and greater_left
  less_right[] <- 0L
  greater_right[] <- 0L


  # -------- Left sweep (groups processed from left to right) ----------
  bit[] <- 0L   # reset BIT in-place (no new allocation)
  total_seen <- 0L
  for (g in 1L:G) {
    i1 <- group_starts[g]
    i2 <- group_ends[g]
    # query each index in this group (do NOT add them yet)
    for (i in i1:i2) {
      r <- ry[i]
      # less_left: bit_sum(r-1)
      j <- r - 1L
      s_less <- 0L
      while (j > 0L) {
        s_less <- s_less + bit[j]
        j <- j - lowbits[j] #bitwAnd(j, -j)
      }
      # leq_count: bit_sum(r)
      j <- r
      s_leq <- 0L
      while (j > 0L) {
        s_leq <- s_leq + bit[j]
        j <- j - lowbits[j] #bitwAnd(j, -j)
      }
      # less_left[i] <- s_less
      # greater_left[i] <- total_seen - s_leq
      less_right[i] <- s_less
      greater_right[i] <- total_seen - s_leq
    }
    # now add this group's ranks to BIT
    for (i in i1:i2) {
      r <- ry[i]
      j <- r
      while (j <= m) {
        bit[j] <- bit[j] + 1L
        j <- j + lowbits[j] #bitwAnd(j, -j)
      }
      total_seen <- total_seen + 1L
    }
  }

  out_xorder <- out_xorder + (less_right - greater_right)
  # reuse some memory
  less_right[] <- 0L
  less_right[ord] <- out_xorder
  return(less_right)

  # combine contributions (in x-sorted order), then restore original order
  # out_xorder <- (greater_right - less_right) + (less_left - greater_left)
  # out <- integer(n)
  # out[ord] <- out_xorder
  # out
}

File: src/concordance.cpp
#include <cpp11.hpp>  // doubles, integers, writable::integers
#include <vector>     // std::vector
#include <algorithm>  // std::stable_sort, std::sort
#include <numeric>    // std::iota
#include <cstring>    // std::memset

using namespace cpp11;

static inline int signum(double x) {
	return (x > 0) - (x < 0);
}

// Naive O(n^2) concordance: sign(xj - xi) * sign(yj - yi) for each i
[[cpp11::register]]
integers concordance_naive_cpp(doubles x, doubles y) {
	int n = x.size();
	writable::integers out(n);

	for (int i = 0; i < n; i++) {
		int s = 0;
		for (int j = 0; j < n; j++)
			if (i != j)
				s += signum(x[j] - x[i]) * signum(y[j] - y[i]);
		out[i] = s;
	}
	return out;
}

// O(n log n) Fenwick-tree concordance
[[cpp11::register]]
integers concordance_fenwick_cpp(doubles x, doubles y) {
	int n = x.size();
	if (n == 0) return writable::integers(static_cast<R_xlen_t>(0));
	if (n == 1) return writable::integers({0});

	// 1) order by x (stable), breaking ties by y
	std::vector<int> ord(n);
	std::iota(ord.begin(), ord.end(), 0);
	std::stable_sort(ord.begin(), ord.end(), [&](int a, int b) {
		return x[a] != x[b] ? x[a] < x[b] : y[a] < y[b];
	});

	// 2) rank-compress y to 1..m via indirect sort (no copies of x/y needed)
	std::vector<int> y_order(n);
	std::iota(y_order.begin(), y_order.end(), 0);
	std::sort(y_order.begin(), y_order.end(), [&](int a, int b) {
		return y[ord[a]] < y[ord[b]];
	});

	std::vector<int> ry(n);
	int m = 0;
	for (int i = 0; i < n; i++) {
		if (i == 0 || y[ord[y_order[i]]] != y[ord[y_order[i - 1]]])
			m++;
		ry[y_order[i]] = m;
	}

	// Fenwick tree (1-indexed) and frequency array to avoid a second BIT query
	std::vector<int> bit(m + 1, 0);
	std::vector<int> freq(m + 1, 0);
	std::vector<int> result(n, 0);

	auto update = [&](int i, int d) {
		for (; i <= m; i += i & (-i)) bit[i] += d;
	};
	auto query = [&](int i) {
		int s = 0;
		for (; i > 0; i -= i & (-i)) s += bit[i];
		return s;
	};

	// ---- Right sweep (groups right to left) ----
	int total = 0, g_end = n - 1;
	while (g_end >= 0) {
		int g_start = g_end;
		while (g_start > 0 && x[ord[g_start - 1]] == x[ord[g_end]])
			g_start--;

		for (int i = g_start; i <= g_end; i++) {
			int r    = ry[i];
			int less = query(r - 1);
			result[i] = (total - less - freq[r]) - less;
		}
		for (int i = g_start; i <= g_end; i++) {
			update(ry[i], 1);
			freq[ry[i]]++;
			total++;
		}
		g_end = g_start - 1;
	}

	// Reset
	std::memset(bit.data(),  0, (m + 1) * sizeof(int));
	std::memset(freq.data(), 0, (m + 1) * sizeof(int));
	total = 0;

	// ---- Left sweep (groups left to right) ----
	int g_start = 0;
	while (g_start < n) {
		int ge = g_start;
		while (ge < n - 1 && x[ord[ge + 1]] == x[ord[g_start]])
			ge++;

		for (int i = g_start; i <= ge; i++) {
			int r    = ry[i];
			int less = query(r - 1);
			result[i] += less - (total - less - freq[r]);
		}
		for (int i = g_start; i <= ge; i++) {
			update(ry[i], 1);
			freq[ry[i]]++;
			total++;
		}
		g_start = ge + 1;
	}

	// Restore original order
	writable::integers out(n);
	for (int i = 0; i < n; i++)
		out[ord[i]] = result[i];
	return out;
}
